README
Conreality Webhooks
===================

See <https://github.com/conreality/conreality-webhooks> for documentation.

Use <https://github.com/conreality/conreality-webhooks/issues> for bug reports.


CHANGES.rst
*********
Changelog
*********

All notable changes to this project will be documented in this file.

The format is based on `Keep a Changelog
<https://keepachangelog.com/en/1.0.0/>`__ and this project adheres to
`Semantic Versioning <https://semver.org/spec/v2.0.0.html>`__.

2019-02-05
==========

Changed
-------

- Rewrote the GitHub webhook in Go for AWS Lambda.


github.php
<?php
/**
 * Post-push webhook for project GitHub repositories.
 *
 * See: https://github.com/conreality/prov.conreality.org/settings/hooks
 * See: https://twitter.com/ConrealityProv
 * See: https://developer.github.com/webhooks/
 * See: https://support.twitter.com/articles/78124
 */

require __DIR__ . '/vendor/autoload.php';

function fail_and_die($reason) {
  http_response_code(500);
  header('Content-Type: text/plain; charset=UTF-8');
  die("500 Internal Server Error ($reason)");
}

$secret_file = __DIR__ . '/../../.secret/twitter.conrealityprov.json';
if (!file_exists($secret_file)) {
  fail_and_die("Missing credentials file");
}

$secret = json_decode(file_get_contents($secret_file));
if (empty($secret->consumer_key) || empty($secret->consumer_secret) ||
    empty($secret->access_token) || empty($secret->access_token_tecret)) {
  fail_and_die("Missing credentials data");
}

if (@$_SERVER['HTTP_X_GITHUB_EVENT'] != 'push') {
  http_response_code(200);
  header('Content-Type: text/plain; charset=UTF-8');
  die();
}

$push = file_get_contents('php://input');
//file_put_contents('.push.json', $push);  // DEBUG
//$push = file_get_contents('.push.json'); // DEBUG
$push = json_decode($push);

$repository_name = $push->repository->name;
$pusher_name     = $push->pusher->name; // good enough for now
$commit          = $push->head_commit;
$commit_sha1     = $commit->id;
$commit_url      = $commit->url;
$commit_text     = $commit->message;

if (strpos($commit_text, "\n") !== false) {
  $commit_text = substr($commit_text, 0, strpos($commit_text, "\n"));
}

if (strlen($commit_text) > 58) { // 40+1+2+1+12+2+58+1+23 = 140
  $commit_text = substr($commit_text, 0, 58 - 1) . "\u{2026}";
}

$log_message =  "$commit_sha1 by $pusher_name: $commit_text\n$commit_url";
//var_dump($log_message); die(); // DEBUG

$twitter = new Twitter($secret->consumer_key, $secret->consumer_secret,
  $secret->access_token, $secret->access_token_tecret);

try {
  $twitter->send($log_message);
}
catch (TwitterException $error) {
  fail_and_die('Twitter API: ' . $error->getMessage());
}

http_response_code(202);
header('Content-Type: text/plain; charset=UTF-8');
print($log_message);


github_webhook.go
/* This is free and unencumbered software released into the public domain. */

// github_webhook posts commits to https://twitter.com/ConrealityCode
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"

	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/dghubble/go-twitter/twitter"
	"github.com/dghubble/oauth1"
	"github.com/pkg/errors"
)

const maxTweetLength = 280

type PingRequest struct {
	// no fields of importance
}

type PushRequest struct {
	HeadCommit PushCommit   `json:"head_commit"`
	Commits    []PushCommit `json:"commits"`
}

type PushCommit struct {
	ID        string    `json:"id"`
	URL       string    `json:"url"`
	Message   string    `json:"message"`
	Author    GitAuthor `json:"author"`
	Committer GitAuthor `json:"committer"`
}

type GitAuthor struct {
	Name  string `json:"name"`
	Email string `json:"email"`
}

func parsePingRequest(request events.APIGatewayProxyRequest) (*PingRequest, error) {
	var payload PingRequest
	err := json.Unmarshal([]byte(request.Body), &payload)
	if err != nil {
		return nil, errors.Wrap(err, "parsePingRequest failed")
	}
	return &payload, nil
}

func parsePushRequest(request events.APIGatewayProxyRequest) (*PushRequest, error) {
	var payload PushRequest
	err := json.Unmarshal([]byte(request.Body), &payload)
	if err != nil {
		return nil, errors.Wrap(err, "parsePushRequest failed")
	}
	return &payload, nil
}

func postTweet(tweetBody string) error {
	config := oauth1.NewConfig(os.Getenv("TWITTER_CONSUMER_KEY"), os.Getenv("TWITTER_CONSUMER_SECRET"))
	token := oauth1.NewToken(os.Getenv("TWITTER_ACCESS_TOKEN"), os.Getenv("TWITTER_ACCESS_SECRET"))
	httpClient := config.Client(oauth1.NoContext, token)
	client := twitter.NewClient(httpClient)
	_, _, err := client.Statuses.Update(tweetBody, nil)
	if err != nil {
		return errors.Wrap(err, "postTweet failed")
	}
	return nil
}

func handleRequest(request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
	event := request.Headers["X-GitHub-Event"]

	log.Print("Handling request with X-GitHub-Event=", event)
	log.Print(request)

	if event == "ping" {
		payload, err := parsePingRequest(request)
		log.Print(payload)
		if err != nil {
			return events.APIGatewayProxyResponse{
				StatusCode: 500,
				Headers:    map[string]string{"content-type": "text/plain"},
				Body:       fmt.Sprintf("Failed to parse a ping request:\n%s\n", err),
			}, nil
		}
		return events.APIGatewayProxyResponse{Body: "OK\n", StatusCode: 200}, nil
	}

	if event == "push" {
		payload, err := parsePushRequest(request)
		if err != nil {
			return events.APIGatewayProxyResponse{
				StatusCode: 500,
				Headers:    map[string]string{"content-type": "text/plain"},
				Body:       fmt.Sprintf("Failed to parse a push request:\n%s\n", err),
			}, nil
		}
		commit := payload.HeadCommit
		commitMessage := commit.Message
		fullMessage := []rune(commitMessage)
		maxMessageLength := maxTweetLength - len(commit.ID) - len(commit.Author.Name) - len(commit.URL) - 8
		if len(fullMessage) > maxMessageLength {
			commitMessage = string(fullMessage[:maxMessageLength]) + "\u2026"
		}
		tweetBody := fmt.Sprintf("%s by %s: %s\n%s\n", commit.ID, commit.Author.Name, commitMessage, commit.URL)
		log.Printf(tweetBody)
		postTweet(tweetBody)
		return events.APIGatewayProxyResponse{
			StatusCode: 200,
			Headers:    map[string]string{"content-type": "text/plain"},
			Body:       tweetBody,
		}, nil
	}

	return events.APIGatewayProxyResponse{
		StatusCode: 500,
		Headers:    map[string]string{"content-type": "text/plain"},
		Body:       fmt.Sprintf("Failed to grok the X-GitHub-Event header: %s\n", event),
	}, nil
}

func main() {
	lambda.Start(handleRequest)
}
